asllib/Typing.ml

```diff
@@ -642,38 +659,28 @@ module Annotate (C : ANNOTATE_CONFIG) = struct
             (check_type_satisfies loc env t1 t_int)
             (check_type_satisfies loc env t1 t_real)
         in
-        match (Types.get_structure env t1).desc with
-        | T_Int None -> T_Int None |> add_pos_from loc
-        | T_Int (Some cs) ->
+        let t1' = Types.to_well_constrained t1 in
+        match (Types.get_structure env t1').desc with
+        | T_Int UnConstrained -> T_Int UnConstrained |> add_pos_from loc
+        | T_Int (UnderConstrained _) ->
+            assert false (* We used to_well_constrained just before. *)
+        | T_Int (WellConstrained cs) ->
             let neg e = E_Unop (NEG, e) |> add_pos_from e in
             let constraint_minus = function
               | Constraint_Exact e -> Constraint_Exact (neg e)
               | Constraint_Range (top, bot) ->
                   Constraint_Range (neg bot, neg top)
             in
-            T_Int (Some (List.map constraint_minus cs)) |> add_pos_from loc
+            T_Int (WellConstrained (List.map constraint_minus cs))
+            |> add_pos_from loc
         | _ -> (* fail case *) t1)
     | NOT ->
         let+ () = check_structure_bits loc env t1 in
         t1 |: TypingRule.CheckUnop
   (* End *)
 
-    (* Rules:
-       - GNTS: It is illegal for a storage element whose type has the structure
-         of the under-constrained integer to be assigned a value whose type has
-         the structure of the under-constrained integer.
-       - LXQZ: A storage element of type S, where S is any type that does not have the
-         structure of the under-constrained integer type, may only be
-         assigned or initialized with a value of type T if T type-satisfies S
-    *)
-    (* TODO: incomplete. *)
   (* Begin CanAssignTo *)
-  let can_assign_to env s t =
-    let s_struct = Types.get_structure env s
-    and t_struct = Types.get_structure env t in
-    match (s_struct.desc, t_struct.desc) with
-    | T_Int (Some []), T_Int (Some []) -> false
-    | _ -> Types.type_satisfies env t s |: TypingRule.CanAssignTo
+  let can_assign_to env s t = Types.type_satisfies env t s |: TypingRule.CanAssignTo
```

Comment:
Nice. Perhaps we can remove the whole notion of can-assign-to and the related rules?